src/pdm_low_level_f401re.cpp

```diff
@@ -0,0 +1,339 @@
+#if defined(ARDUINO_NUCLEO_F401RE)
+
+#include "PDM.h"
+
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+I2S_HandleTypeDef hi2s3;
+DMA_HandleTypeDef hdma_spi3_rx;
+
+uint16_t I2S_InternalBuffer[((AUDIO_IN_FREQ / 8) * MAX_AUDIO_IN_CHANNEL_NBR_PER_IF * N_MS_PER_INTERRUPT)];
+uint32_t I2S_InternalBufferSize = (((AUDIO_IN_FREQ / 8) * MAX_AUDIO_IN_CHANNEL_NBR_PER_IF *N_MS_PER_INTERRUPT) * 2);
+
+static void MX_DMA_Init(void);
+static int MX_I2S3_Init(void);
+
+/* Low Level Functions implementation */
+
+/**
+ *  @brief PDM initialization
+ *  @param None
+ *  @retval 1 if success, 0 otherwise
+ */
+int pdm_low_level_init(void)
+{
+  int ret = PDM_OK;
+  MX_DMA_Init();
+  ret = MX_I2S3_Init();
+  return ret;
+}
+
+/**
+ *  @brief PDM deinitialization
+ *  @param None
+ *  @retval 1 if success, 0 otherwise
+ */
+int pdm_low_level_deinit(void)
+{
+  if (HAL_I2S_DeInit(&hi2s3) != HAL_OK) {
+    return PDM_ERROR;
+  }
+
+  return PDM_OK;
+}
+
+/**
+ *  @brief Start recording
+ *  @param None
+ *  @retval 1 if success, 0 otherwise
+ */
+int pdm_low_level_record(void)
+{
+  if (HAL_I2S_Receive_DMA(&hi2s3, I2S_InternalBuffer, (uint16_t)I2S_InternalBufferSize / 2U) != HAL_OK) {
+    return PDM_ERROR;
+  }
+
+  return PDM_OK;
+}
+
+/**
+ *  @brief Stop recording
+ *  @param None
+ *  @retval 1 if success, 0 otherwise
+ */
+int pdm_low_level_stop(void)
+{
+  if (HAL_I2S_DMAStop(&hi2s3) != HAL_OK) {
+    return PDM_ERROR;
+  }
+
+  return PDM_OK;
+}
+
+/**
+ *  @brief Pause recording
+ *  @param None
+ *  @retval 1 if success, 0 otherwise
+ */
+int pdm_low_level_pause(void)
+{
+  if (HAL_I2S_DMAPause(&hi2s3) != HAL_OK) {
+    return PDM_ERROR;
+  }
+
+  return PDM_OK;
+}
+
+/**
+ *  @brief Resume recording
+ *  @param None
+ *  @retval 1 if success, 0 otherwise
+ */
+int pdm_low_level_resume(void)
+{
+  if (HAL_I2S_DMAResume(&hi2s3) != HAL_OK) {
+    return PDM_ERROR;
+  }
+
+  return PDM_OK;
+}
+
+/**
+ *  @brief Half Callback
+ *  @param pBuf pointer to a buffer to store data
+ *  @retval None
+ */
+void pdm_low_level_copy_half_cplt(uint16_t *pBuf)
+{
+  uint32_t index;
+  uint16_t *DataTempI2S = I2S_InternalBuffer;
+
+  for (index = 0; index < (I2S_InternalBufferSize / 4U); index++) {
+    pBuf[index] = (DataTempI2S[index]);
+  }
+}
+
+/**
+ *  @brief Complet Callback
+ *  @param pBuf pointer to a buffer to store data
+ *  @retval None
+ */
+void pdm_low_level_copy_cplt(uint16_t *pBuf)
+{
+  uint32_t index;
+  uint16_t *DataTempI2S = &I2S_InternalBuffer[I2S_InternalBufferSize / 4U];
+
+  for (index = 0; index < (I2S_InternalBufferSize / 4U); index++) {
+    pBuf[index] = (DataTempI2S[index]);
+  }
+}
+
+
+/* This function is defined weak in the Arduino Core, so we can override it */
+
+/**
+ *  @brief System Clock Parameters Configuration
+ *  @param None
+ *  @retval None
+ */
+void SystemClock_Config(void)
+{
+  RCC_OscInitTypeDef RCC_OscInitStruct = {0};
+  RCC_ClkInitTypeDef RCC_ClkInitStruct = {0};
```

Comment:
Hi @sarachiodo ,
I noticed that the Arduino compiler gives a warning. We need to replace {0} with simply {}